pkg/k8s/statedb_test.go

```diff
@@ -0,0 +1,252 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/k8s"
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/utils"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+func ExampleRegisterReflector() {
+	var module = cell.Module(
+		"example-reflector",
+		"Reflector example",
+
+		cell.ProvidePrivate(
+			// The table the reflector writes to (RWTable[*Node]).
+			newTestNodeTable,
+
+			// ReflectorConfig defines the ListerWatcher to use the fetch the objects
+			// and how to write them to the table.
+			func(client k8sClient.Clientset) k8s.ReflectorConfig[*corev1.Node] {
+				return k8s.ReflectorConfig[*corev1.Node]{
+					ListerWatcher: utils.ListerWatcherFromTyped(client.CoreV1().Nodes()),
+				}
+			},
+		),
+
+		// Provide Table[*Node] for read-access to all modules in the application.
+		cell.Provide(statedb.RWTable[*corev1.Node].ToTable),
+
+		// Register the reflector to this module's job group.
+		cell.Invoke(k8s.RegisterReflector[*corev1.Node]),
+	)
+
+	hive.New(module)
+}
+
+var (
+	testNodeNameIndex = statedb.Index[*corev1.Node, string]{
+		Name: "name",
+		FromObject: func(obj *corev1.Node) index.KeySet {
+			return index.NewKeySet(index.String(obj.Name))
+		},
+		FromKey: index.String,
+		Unique:  true,
+	}
+)
+
+func newTestNodeTable(db *statedb.DB) (statedb.RWTable[*corev1.Node], error) {
+	tbl, err := statedb.NewTable(
+		"test-nodes",
+		testNodeNameIndex,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return tbl, db.RegisterTable(tbl)
+}
+
+func TestReflector(t *testing.T) {
+	testReflector(t, false, false)
+}
+
+func TestReflector_TransformQueryAll(t *testing.T) {
+	testReflector(t, true, true)
+}
+
+func testReflector(t *testing.T, doTransform, doQueryAll bool) {
+	var (
+		node1Name = "node1"
+		node2Name = "node2"
+		node      = &corev1.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            node1Name,
+				ResourceVersion: "0",
+			},
+			Status: corev1.NodeStatus{
+				Phase: "init",
+			},
+		}
+		fakeClient, cs = k8sClient.NewFakeClientset()
+
+		db                              *statedb.DB
+		nodeTable                       statedb.Table[*corev1.Node]
+		transformCalled, queryAllCalled atomic.Bool
+	)
+
+	// Create the initial version of the node. Do this before anything
+	// starts watching the resources to avoid a race.
+	fakeClient.KubernetesFakeClientset.Tracker().Create(
+		corev1.SchemeGroupVersion.WithResource("nodes"),
+		node.DeepCopy(), "")
+
+	var testTimeout = 10 * time.Second
+	ctx, cancel := context.WithTimeout(context.Background(), testTimeout)
+	defer cancel()
+
+	var transformFunc k8s.TransformFunc[*corev1.Node]
+	if doTransform {
+		transformFunc = func(a any) (obj *corev1.Node, ok bool) {
+			transformCalled.Store(true)
+			obj = a.(*corev1.Node).DeepCopy()
+			obj.ObjectMeta.GenerateName = "transformed"
+			return obj, true
+		}
+	}
+
+	var queryAllFunc k8s.QueryAllFunc[*corev1.Node]
```

Comment:
Nit: Lint suggests to merge this line with the next one.

Reply:
oopsie, forgot the "if doQueryAll" (want to test both with a nil and non-nil func here). fixed.